depth_to_pointcloud/src/depth_to_pointcloud_node.cpp

```diff
@@ -0,0 +1,180 @@
+
+  sensor_msgs::PointCloud2Iterator<float> iter_x(*cloud_msg, "x");
+  sensor_msgs::PointCloud2Iterator<float> iter_y(*cloud_msg, "y");
+  sensor_msgs::PointCloud2Iterator<float> iter_z(*cloud_msg, "z");
+  const T * depth_row = reinterpret_cast<const T *>(&depth_msg->data[0]);
+  int row_step = depth_msg->step / sizeof(T);
+  for (int v = 0; v < static_cast<int>(cloud_msg->height); ++v, depth_row += row_step) {
+    for (int u = 0; u < static_cast<int>(cloud_msg->width); ++u, ++iter_x, ++iter_y, ++iter_z) {
+      T depth = depth_row[u];
+
+      // Missing points denoted by NaNs
+      if (!DepthTraits<T>::valid(depth)) {
+        if (range_max != 0.0) {
+          depth = DepthTraits<T>::fromMeters(range_max);
+        } else {
+          *iter_x = *iter_y = *iter_z = bad_point;
+          continue;
+        }
+      }
+
+      // Fill in XYZ
+      *iter_x = (u - center_x) * depth * constant_x;
+      *iter_y = (v - center_y) * depth * constant_y;
+      *iter_z = DepthTraits<T>::toMeters(depth);
+    }
+  }
+}
+
+static void depthCb(const sensor_msgs::msg::Image::SharedPtr image)
+{
+  // The meat of this function is a port of the code from:
+  // https://github.com/ros-perception/image_pipeline/blob/indigo/depth_image_proc/src/nodelets/point_cloud_xyz.cpp
+
+  if (g_cam_info == nullptr) {
+    // we haven't gotten the camera info yet, so just drop until
+    // we do.
+    fprintf(stderr, "No camera info, skipping point cloud conversion\n");
+    return;
+  }
+  sensor_msgs::msg::PointCloud2::SharedPtr cloud_msg =
+    std::make_shared<sensor_msgs::msg::PointCloud2>();
+  cloud_msg->header = image->header;
+  cloud_msg->height = image->height;
+  cloud_msg->width = image->width;
+  cloud_msg->is_dense = false;
+  cloud_msg->is_bigendian = false;
+  cloud_msg->fields.clear();
+  cloud_msg->fields.reserve(1);
+
+  sensor_msgs::PointCloud2Modifier pcd_modifier(*cloud_msg);
+  pcd_modifier.setPointCloud2FieldsByString(1, "xyz");
+
+  // g_cam_info here is a sensor_msg::msg::CameraInfo::ConstSharedPtr,
+  // which we get from the cam_info topic.
+  image_geometry::PinholeCameraModel model;
+  model.fromCameraInfo(g_cam_info);
+
+  if (image->encoding == sensor_msgs::image_encodings::TYPE_16UC1) {
+    convert<uint16_t>(image, cloud_msg, model);
+  } else if (image->encoding == sensor_msgs::image_encodings::TYPE_32FC1) {
+    convert<float>(image, cloud_msg, model);
+  } else {
+    fprintf(stderr, "Depth image has unsupported encoding [%s]\n", image->encoding.c_str());
+    return;
+  }
+
+  g_pub_point_cloud->publish(cloud_msg);
+}
+
+void infoCb(sensor_msgs::msg::CameraInfo::SharedPtr info)
+{
+  g_cam_info = info;
+}
+
+int main(int argc, char ** argv)
+{
+  rclcpp::init(argc, argv);
+
+  rclcpp::node::Node::SharedPtr node = rclcpp::node::Node::make_shared("depth_to_pointcloud");
+
+  rmw_qos_profile_t custom_qos_profile = rmw_qos_profile_default;
+
+  custom_qos_profile.depth = 1;
```

Comment:
Remark: Not sure you need the same qos profile for the camera info and the images though, they deal with very different message size. This can be modified after this PR when needed.